activerecord/lib/active_record/locking/optimistic.rb

```diff
@@ -85,10 +85,14 @@ def _update_row(attribute_names, attempted_action = "update")
 
             self[locking_column] += 1
 
-            affected_rows = self.class._update_record(
-              attributes_with_values(attribute_names),
-              @primary_key => id_in_database,
-              locking_column => previous_lock_value
+            # When `lock_version` is read as `0`, it may actually be `NULL` in the DB.
+            possible_previous_lock_value = previous_lock_value.to_i == 0 ? [nil, 0] : previous_lock_value
+
+            affected_rows = self.class.unscoped.where(
+              locking_column => possible_previous_lock_value,
+              self.class.primary_key => id_in_database
+            ).update_all(
+              attributes_with_values(attribute_names)
```

Comment:
By internal design, we should use `_update_record` here, I guess we should make `_update_record` understand how to build constraint from an array value:
```ruby
self.class._update_record(
locking_column => [nil, 0] # ("lock_version" IS NULL OR "lock_version" = "0")
...
)
```

/cc @kamipo

Reply:
Would you suggest adding a case to `_update_record` (and the other `_*record` methods) to deal with this? Something like:

```
        constraints = _substitute_values(constraints).map do |attr, bind|
          if bind.is_a?(Enumerable)
            bind.map { |bind_value| attr.eq(bind_value) }.reduce(&:or)
          else
            attr.eq(bind)
          end
        end
```

though this would also require changing `_subtitute_values` to something like

```
            bind = if value.is_a?(Enumerable)
              value.map { |single_value| predicate_builder.build_bind_attribute(name, single_value) }
            else
              predicate_builder.build_bind_attribute(name, value)
            end
```

WDYT?

Reply:
Ah, we could actually use the `ActiveRecord::PredicateBuilder::ArrayHandler` in the `_(update|delete)_record` methods like so (but would still need to change `_substitute_values` as above):

```
        constraints = _substitute_values(constraints).map do |attr, bind|
          if bind.is_a?(Enumerable)
            ActiveRecord::PredicateBuilder::ArrayHandler.new(predicate_builder).call(attr, bind)
          else
            attr.eq(bind)
          end
        end
```

Reply:
Is there a reason why the predicate is being built directly inside `_(update|delete)_record` btw? Why isn't `predicate_builder#build` used?

Reply:
@bogdanvlviv sorry to ping you again, just wondering if you had any insights from my questions above?

Reply:
First of all, I want to apologize for the unclear first comment. By "By internal design, we should use _update_record here." I meant that it would be great to retain using `_update_record` for consistency, also using this method should get better performance(bbae710a405ce92074c1666dcf859196c29e2ed2) since it looks like in this case we prevent building AR-relation.

I've spent some time looking at this, but unfortunately, I haven't got any good answers to your questions yet.

Actually, I am fine with this implementation since it eliminates the issue. I just want to know @kamipo's thoughts about it. Related to fa779b380e61381a393afbc7bbc0a9ce07e0ce74

Reply:
Sorry to be a bother - it's been quite a bit (over a month now 😅) and was wondering if we could maybe get someone else's thoughts about it instead, since @kamipo seems to be quite busy?

Reply:
`"posts"."lock_version" IS NULL OR "posts"."lock_version" = "0"` is redundant, it should just be to be `"posts"."lock_version" IS NULL` if `lock_version` was nil.

Reply:
@kamipo thanks for taking a look! The problem is that we cannot know if `lock_version` was `nil` or `0` - it's unknown to us.

`previous_lock_value = read_attribute_before_type_cast(locking_column)` will hold a `0` even if the lock version is `nil`

Reply:
so because of that, we must use `"posts"."lock_version" IS NULL OR "posts"."lock_version" = "0"` as we can't know what the actual value is in the DB

I could of course be missing something, could you please expand on your comment @kamipo ?

Reply:
sorry for pinging yet again @kamipo -  I left some comments here on why we can't really do what you're suggesting. Could you please take a look?